pkg/admissionpolicy/mutate.go

```diff
@@ -112,23 +126,310 @@ func mutateResource(
+
+		serverGVR, err := client.Discovery().GetGVRFromGVK(gvk)
+		if err != nil {
+			return emptyResp, err
+		}
+
+		a = admission.NewAttributesRecord(
+			resource.DeepCopyObject(), nil,
+			gvk,
+			resource.GetNamespace(), resource.GetName(),
+			serverGVR, "", admission.Create, nil, false, nil,
+		)
+
+		o := admission.NewObjectInterfacesFromScheme(runtime.NewScheme())
+
+		// match policy
+		isPolicyMatch, _, _, err := matcher.DefinitionMatches(a, o, NewMutatingAdmissionPolicyAccessor(&policy))
+		if err != nil {
+			return emptyResp, err
+		}
+		if !isPolicyMatch {
+			return emptyResp, nil
+		}
+
+		// match bindings
+		for i, binding := range bindings {
+			isBindingMatch, err := matcher.BindingMatches(a, o, NewMutatingAdmissionPolicyBindingAccessor(&binding))
+			if err != nil {
+				return emptyResp, err
+			}
+			if !isBindingMatch {
+				continue
+			}
+
+			logger.V(3).Info("mutate resource %s against policy %s with binding %s", resPath, policy.GetName(), binding.GetName())
+			return MutateResource(policy, &bindings[i], resource)
+		}
+
+		return emptyResp, nil
+	} else {
+		// client == nil → offline matching
+		isPolicyMatch, err := matches(a, namespaceSelectorMap, convertMatchResources(*policy.Spec.MatchConstraints))
+		if err != nil {
+			return emptyResp, err
+		}
+		if !isPolicyMatch {
+			return emptyResp, nil
+		}
+
+		for i, binding := range bindings {
+			isBindingMatch, err := matches(a, namespaceSelectorMap, convertMatchResources(*binding.Spec.MatchResources))
+			if err != nil {
+				return emptyResp, err
+			}
+			if !isBindingMatch {
+				continue
+			}
+
+			logger.V(3).Info("mutate resource %s against policy %s with binding %s", resPath, policy.GetName(), binding.GetName())
+			return MutateResource(policy, &bindings[i], resource)
+		}
+
+		return emptyResp, nil
+	}
+}
+func NewMutatingAdmissionPolicyBindingAccessor(binding *admissionregistrationv1alpha1.MutatingAdmissionPolicyBinding) generic.BindingAccessor {
+	return &mutatingBindingAccessor{binding}
+}
+
+type mutatingBindingAccessor struct {
+	binding *admissionregistrationv1alpha1.MutatingAdmissionPolicyBinding
+}
+
+func (b *mutatingBindingAccessor) GetName() string {
+	return b.binding.Name
+}
+
+func (b *mutatingBindingAccessor) GetNamespace() string {
+	return b.binding.Namespace
+}
+
+func (b *mutatingBindingAccessor) GetMatchResources() *admissionregistrationv1.MatchResources {
+	if b.binding.Spec.MatchResources == nil {
+		return nil
+	}
+	converted := convertMatchResources(*b.binding.Spec.MatchResources)
+	return &converted
+}
+func (b *mutatingBindingAccessor) GetParamRef() *admissionregistrationv1.ParamRef {
+	return nil
+}
+
+func (b *mutatingBindingAccessor) GetPolicyName() types.NamespacedName {
+	return types.NamespacedName{
+		// no explicit policy field, so we can only provide binding's own namespace and empty name
+		Namespace: b.binding.Namespace,
+		Name:      "", // <-- no policy name, because field is missing
+	}
+}
```

Comment:
These functions already exist in the upstream code [here](https://github.com/kubernetes/apiserver/blob/master/pkg/admission/plugin/policy/mutating/accessor.go).
You need to import it like this:
```
import (
    "k8s.io/apiserver/pkg/admission/plugin/policy/mutating"
)
```
And to use it:
```
 mutating.NewMutatingAdmissionPolicyAccessor(&policy)
```